storage/cleanup.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/m3db/m3x/errors"
+)
+
+type filesetFilesBeforeFn func(filePathPrefix string, shardID uint32, t time.Time) ([]string, error)
+
+type commitLogFilesBeforeFn func(commitLogsDir string, t time.Time) ([]string, error)
+
+type commitLogFilesForTimeFn func(commitLogsDir string, t time.Time) ([]string, error)
+
+type deleteFilesFn func(files []string) error
+
+type cleanupManager struct {
+	database                database
+	opts                    Options
+	blockSize               time.Duration
+	filePathPrefix          string
+	commitLogsDir           string
+	fm                      databaseFlushManager
+	filesetFilesBeforeFn    filesetFilesBeforeFn
+	commitLogFilesBeforeFn  commitLogFilesBeforeFn
+	commitLogFilesForTimeFn commitLogFilesForTimeFn
+	deleteFilesFn           deleteFilesFn
+}
+
+func newCleanupManager(database database, fm databaseFlushManager) databaseCleanupManager {
+	opts := database.Options()
+	blockSize := opts.GetRetentionOptions().GetBlockSize()
+	filePathPrefix := opts.GetCommitLogOptions().GetFilesystemOptions().GetFilePathPrefix()
+	commitLogsDir := fs.CommitLogsDirPath(filePathPrefix)
+
+	return &cleanupManager{
+		database:                database,
+		opts:                    opts,
+		blockSize:               blockSize,
+		filePathPrefix:          filePathPrefix,
+		commitLogsDir:           commitLogsDir,
+		fm:                      fm,
+		filesetFilesBeforeFn:    fs.FilesetFilesBefore,
+		commitLogFilesBeforeFn:  fs.CommitLogFilesBefore,
+		commitLogFilesForTimeFn: fs.CommitLogFilesForTime,
+		deleteFilesFn:           fs.DeleteFiles,
+	}
+}
+
+func (mgr *cleanupManager) Cleanup(t time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	filesetFilesStart := mgr.fm.FlushTimeStart(t)
+	if err := mgr.cleanupFilesetFiles(filesetFilesStart); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up fileset files for %v: %v", filesetFilesStart, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	commitLogStart, commitLogTimes := mgr.commitLogTimes(t)
+	if err := mgr.cleanupCommitLogs(commitLogStart, commitLogTimes); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up commit logs for commitLogStart %v commitLogTimes %v: %v", commitLogStart, commitLogTimes, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	return multiErr.FinalError()
+}
+
+func (mgr *cleanupManager) cleanupFilesetFiles(earliestToRetain time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	shards := mgr.database.getOwnedShards()
+	for _, shard := range shards {
+		shardID := shard.ShardNum()
+		expired, err := mgr.filesetFilesBeforeFn(mgr.filePathPrefix, shardID, earliestToRetain)
+		if err != nil {
+			detailedErr := fmt.Errorf("encountered errors when getting fileset files for prefix %s shard %d: %v", mgr.filePathPrefix, shardID, err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+		if err := mgr.deleteFilesFn(expired); err != nil {
+			multiErr = multiErr.Add(err)
```

Comment:
The error returned from `fs.DeleteFiles` already has all the information including the files that caused the error and the actual error.